String datatype coordinates fail in QC `CheckMonotonic`

This shared QC test is automatically set to run all coordinates in pipeline template. Occasionally (often?) 2 or 3D variables contain a frame of reference coordinate with a string datatype (XYZ, ENU, etc), and these variables will error out in CheckMonotonic. The short-term solution is to specify non-string variables for that particular test.

Reproducible example:

```python
import numpy as np
import pandas as pd
import xarray as xr

from tsdat import CheckMonotonic

ds = xr.Dataset(
    coords={
        "time": pd.date_range("2022-03-24 21:43:00", "2022-03-24 21:45:00", periods=3),  # type: ignore
        "dir": ["N", "E", "S", "W"],
    },
    data_vars={
        "wind_speed": (["time", "dir"], np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]))  # type: ignore
    },
)
# expected = np.array(3 * [4 * [False]])
expected = np.array([False, False, False, False])
results = CheckMonotonic().run(ds, "dir")  # type: ignore
assert np.array_equal(results, expected)  # type: ignore
```

<details>
<summary>Trackback message</summary>

```
./test/qc/test_qc.py::test_monotonic_check_ignores_string_vars Failed: [undefined]numpy.core._exceptions._UFuncNoLoopError: ufunc 'subtract' did not contain a loop with signature matching types (dtype('<U1'), dtype('<U1')) -> None
def test_monotonic_check_ignores_string_vars():
        ds = xr.Dataset(
            coords={
                "time": pd.date_range("2022-03-24 21:43:00", "2022-03-24 21:45:00", periods=3),  # type: ignore
                "dir": ["N", "E", "S", "W"],
            },
            data_vars={
                "wind_speed": (["time", "dir"], np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]))  # type: ignore
            },
        )
        # expected = np.array(3 * [4 * [False]])
        expected = np.array([False, False, False, False])
>       results = CheckMonotonic().run(ds, "dir")  # type: ignore

test/qc/test_qc.py:188: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tsdat/qc/checkers.py:83: in run
    diff: NDArray[Any] = np.diff(variable.data, axis=axis)  # type: ignore
<__array_function__ internals>:180: in diff
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = array(['N', 'E', 'S', 'W'], dtype='<U1'), n = 1, axis = 0
prepend = <no value>, append = <no value>

    @array_function_dispatch(_diff_dispatcher)
    def diff(a, n=1, axis=-1, prepend=np._NoValue, append=np._NoValue):
        """
        Calculate the n-th discrete difference along the given axis.
    
        The first difference is given by ``out[i] = a[i+1] - a[i]`` along
        the given axis, higher differences are calculated by using `diff`
        recursively.
    
        Parameters
        ----------
        a : array_like
            Input array
        n : int, optional
            The number of times values are differenced. If zero, the input
            is returned as-is.
        axis : int, optional
            The axis along which the difference is taken, default is the
            last axis.
        prepend, append : array_like, optional
            Values to prepend or append to `a` along axis prior to
            performing the difference.  Scalar values are expanded to
            arrays with length 1 in the direction of axis and the shape
            of the input array in along all other axes.  Otherwise the
            dimension and shape must match `a` except along axis.
    
            .. versionadded:: 1.16.0
    
        Returns
        -------
        diff : ndarray
            The n-th differences. The shape of the output is the same as `a`
            except along `axis` where the dimension is smaller by `n`. The
            type of the output is the same as the type of the difference
            between any two elements of `a`. This is the same as the type of
            `a` in most cases. A notable exception is `datetime64`, which
            results in a `timedelta64` output array.
    
        See Also
        --------
        gradient, ediff1d, cumsum
    
        Notes
        -----
        Type is preserved for boolean arrays, so the result will contain
        `False` when consecutive elements are the same and `True` when they
        differ.
    
        For unsigned integer arrays, the results will also be unsigned. This
        should not be surprising, as the result is consistent with
        calculating the difference directly:
    
        >>> u8_arr = np.array([1, 0], dtype=np.uint8)
        >>> np.diff(u8_arr)
        array([255], dtype=uint8)
        >>> u8_arr[1,...] - u8_arr[0,...]
        255
    
        If this is not desirable, then the array should be cast to a larger
        integer type first:
    
        >>> i16_arr = u8_arr.astype(np.int16)
        >>> np.diff(i16_arr)
        array([-1], dtype=int16)
    
        Examples
        --------
        >>> x = np.array([1, 2, 4, 7, 0])
        >>> np.diff(x)
        array([ 1,  2,  3, -7])
        >>> np.diff(x, n=2)
        array([  1,   1, -10])
    
        >>> x = np.array([[1, 3, 6, 10], [0, 5, 6, 8]])
        >>> np.diff(x)
        array([[2, 3, 4],
               [5, 1, 2]])
        >>> np.diff(x, axis=0)
        array([[-1,  2,  0, -2]])
    
        >>> x = np.arange('1066-10-13', '1066-10-16', dtype=np.datetime64)
        >>> np.diff(x)
        array([1, 1], dtype='timedelta64[D]')
    
        """
        if n == 0:
            return a
        if n < 0:
            raise ValueError(
                "order must be non-negative but got " + repr(n))
    
        a = asanyarray(a)
        nd = a.ndim
        if nd == 0:
            raise ValueError("diff requires input that is at least one dimensional")
        axis = normalize_axis_index(axis, nd)
    
        combined = []
        if prepend is not np._NoValue:
            prepend = np.asanyarray(prepend)
            if prepend.ndim == 0:
                shape = list(a.shape)
                shape[axis] = 1
                prepend = np.broadcast_to(prepend, tuple(shape))
            combined.append(prepend)
    
        combined.append(a)
    
        if append is not np._NoValue:
            append = np.asanyarray(append)
            if append.ndim == 0:
                shape = list(a.shape)
                shape[axis] = 1
                append = np.broadcast_to(append, tuple(shape))
            combined.append(append)
    
        if len(combined) > 1:
            a = np.concatenate(combined, axis)
    
        slice1 = [slice(None)] * nd
        slice2 = [slice(None)] * nd
        slice1[axis] = slice(1, None)
        slice2[axis] = slice(None, -1)
        slice1 = tuple(slice1)
        slice2 = tuple(slice2)
    
        op = not_equal if a.dtype == np.bool_ else subtract
        for _ in range(n):
>           a = op(a[slice1], a[slice2])
E           numpy.core._exceptions._UFuncNoLoopError: ufunc 'subtract' did not contain a loop with signature matching types (dtype('<U1'), dtype('<U1')) -> None

../../anaconda3/envs/tsdat/lib/python3.8/site-packages/numpy/lib/function_base.py:1447: UFuncTypeError
```

</details>

--- a/test/qc/test_qc.py
+++ b/test/qc/test_qc.py
@@ -3,6 +3,7 @@ import xarray as xr
 import pandas as pd
 import numpy as np
 from numpy.typing import NDArray
+import pytest
 from pytest import fixture
 from tsdat.qc.checkers import *
 from tsdat.qc.handlers import *

--- a/test/qc/test_qc.py
+++ b/test/qc/test_qc.py
@@ -158,6 +158,21 @@ def test_check_max_classes(sample_dataset: xr.Dataset):
     assert np.array_equal(results, expected)  # type: ignore
 
 
+def test_valid_delta():
+    ds = xr.Dataset(
+        coords={
+            "time": pd.date_range("2022-03-24 21:43:00", "2022-03-24 21:45:00", periods=3),  # type: ignore
+            "height": np.array([1, 2]),
+        },
+        data_vars={
+            "wind_speed": (["time", "height"], np.array([[10, 20], [11, 25], [16, 31]]), {"units": "m/s", "valid_delta": 5})  # type: ignore
+        },
+    )
+    expected = np.array([[False, False], [False, False], [False, True]])
+    results = CheckValidDelta().run(ds, "wind_speed")
+    assert np.array_equal(results, expected)  # type: ignore
+
+
 def test_check_delta_classes(sample_dataset: xr.Dataset):
     var_name = "monotonic_var"
     expected = np.bool8([False, False, False, True])

--- a/test/qc/test_qc.py
+++ b/test/qc/test_qc.py
@@ -1,10 +1,12 @@
+import logging
 from typing import Any, Dict
-import xarray as xr
-import pandas as pd
+
 import numpy as np
+import pandas as pd
+import xarray as xr
 from numpy.typing import NDArray
-import pytest
 from pytest import fixture
+
 from tsdat.qc.checkers import *
 from tsdat.qc.handlers import *
 from tsdat.testing import assert_close
@@ -174,7 +176,7 @@ def test_valid_delta():
     assert np.array_equal(results, expected)  # type: ignore
 
 
-def test_monotonic_check_ignores_string_vars(capsys: pytest.CaptureFixture[str]):
+def test_monotonic_check_ignores_string_vars(caplog: Any):
     ds = xr.Dataset(
         coords={
             "time": pd.date_range("2022-03-24 21:43:00", "2022-03-24 21:45:00", periods=3),  # type: ignore
@@ -185,8 +187,14 @@ def test_monotonic_check_ignores_string_vars(capsys: pytest.CaptureFixture[str])
         },
     )
     expected = np.array([False, False, False, False])
-    results = CheckMonotonic().run(ds, "dir")  # type: ignore
+
+    with caplog.at_level(logging.WARNING):
+        results = CheckMonotonic().run(ds, "dir")  # type: ignore
     assert np.array_equal(results, expected)  # type: ignore
+    assert (
+        "Variable 'dir' has dtype '<U1', which is currently not supported for monotonicity checks."
+        in caplog.text
+    )
 
 
 def test_check_delta_classes(sample_dataset: xr.Dataset):

--- a/test/qc/test_qc.py
+++ b/test/qc/test_qc.py
@@ -173,6 +173,21 @@ def test_valid_delta():
     assert np.array_equal(results, expected)  # type: ignore
 
 
+def test_monotonic_check_ignores_string_vars(capsys: pytest.CaptureFixture[str]):
+    ds = xr.Dataset(
+        coords={
+            "time": pd.date_range("2022-03-24 21:43:00", "2022-03-24 21:45:00", periods=3),  # type: ignore
+            "dir": ["N", "E", "S", "W"],
+        },
+        data_vars={
+            "wind_speed": (["time", "dir"], np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]))  # type: ignore
+        },
+    )
+    expected = np.array([False, False, False, False])
+    results = CheckMonotonic().run(ds, "dir")  # type: ignore
+    assert np.array_equal(results, expected)  # type: ignore
+
+
 def test_check_delta_classes(sample_dataset: xr.Dataset):
     var_name = "monotonic_var"
     expected = np.bool8([False, False, False, True])

--- a/test/qc/test_qc.py
+++ b/test/qc/test_qc.py
@@ -3,6 +3,7 @@ from typing import Any, Dict
 
 import numpy as np
 import pandas as pd
+import pytest
 import xarray as xr
 from numpy.typing import NDArray
 from pytest import fixture
@@ -263,3 +264,31 @@ def test_sortdataset_by_coordinate(sample_dataset: xr.Dataset):
 
     assert_close(dataset, expected)
     assert dataset.time.attrs.get("corrections_applied") == ["Sorted time data!"]
+
+
+def test_fail_pipeline_provides_useful_message(caplog: Any):
+
+    ds = xr.Dataset(
+        coords={
+            "time": pd.date_range("2022-03-24 21:43:00", "2022-03-24 21:45:00", periods=3),  # type: ignore
+            "dir": ["X", "Y", "Z"],
+        },
+        data_vars={
+            "position": (["time", "dir"], np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))  # type: ignore
+        },
+    )
+
+    failures = np.array(
+        [[False, False, False], [False, False, False], [False, True, False]]
+    )
+    with pytest.raises(
+        DataQualityError, match=r".*Quality results for variable 'position'.*"
+    ) as err:
+        _ = FailPipeline().run(ds, "position", failures)
+
+    msg = err.value.args[0]
+
+    assert "Quality results for variable 'position' indicate a fatal error" in msg
+    assert "1 / 9 values failed" in msg
+    assert "The first failures occur at indexes: [[2, 1]]" in msg
+    assert "The corresponding values are: [8]" in msg

--- a/tsdat/qc/checkers.py
+++ b/tsdat/qc/checkers.py
@@ -350,7 +350,8 @@ class _CheckDelta(_ThresholdChecker):
         data: NDArray[Any] = var_data.data
         axis = var_data.get_axis_num(self.parameters.dim)
 
-        diff: NDArray[Any] = np.absolute(np.diff(data, axis=axis, prepend=data[0]))  # type: ignore
+        prepend = np.expand_dims(np.take(data, 0, axis=axis), axis=axis)  # type: ignore
+        diff: NDArray[Any] = np.absolute(np.diff(data, axis=axis, prepend=prepend))  # type: ignore
         failures = diff > threshold if self.allow_equal else diff >= threshold
 
         return failures

--- a/tsdat/qc/checkers.py
+++ b/tsdat/qc/checkers.py
@@ -1,9 +1,11 @@
+import logging
 import numpy as np
 import xarray as xr
 from pydantic import BaseModel, Extra, validator
 from typing import Any, Dict, List, Optional, Union
 from numpy.typing import NDArray
 from .base import QualityChecker
+
 
 __all__ = [
     "CheckMissing",
@@ -24,6 +26,8 @@ __all__ = [
     "CheckFailDelta",
     "CheckWarnDelta",
 ]
+
+logger = logging.getLogger(__name__)
 
 
 class CheckMissing(QualityChecker):
@@ -77,7 +81,17 @@ class CheckMonotonic(QualityChecker):
     parameters: Parameters = Parameters()
 
     def run(self, dataset: xr.Dataset, variable_name: str) -> NDArray[np.bool8]:
+
         variable = dataset[variable_name]
+
+        if variable.values.dtype.kind in {"U", "S"}:  # type: ignore
+            logger.warning(
+                "Variable '%s' has dtype '%s', which is currently not supported for"
+                " monotonicity checks.",
+                variable_name,
+                variable.values.dtype,  # type: ignore
+            )
+            return np.full(variable.shape, False, dtype=np.bool8)
 
         axis = self.get_axis(variable)
         diff: NDArray[Any] = np.diff(variable.data, axis=axis)  # type: ignore

--- a/tsdat/qc/handlers.py
+++ b/tsdat/qc/handlers.py
@@ -1,11 +1,12 @@
+from typing import Any, List, Literal, Tuple, Union
+
 import numpy as np
 import xarray as xr
-from pydantic import BaseModel, Extra, Field, validator
-from typing import Any, Literal
 from numpy.typing import NDArray
-from .base import QualityHandler
-from ..utils import record_corrections_applied
+from pydantic import BaseModel, Extra, Field, validator
 
+from ..utils import record_corrections_applied
+from .base import QualityHandler
 
 __all__ = [
     "DataQualityError",
@@ -41,21 +42,53 @@ class FailPipeline(QualityHandler):
         context: str = ""
         """Additional context set by users that ends up in the traceback message."""
 
+        display_limit: int = 5
+
     parameters: Parameters = Parameters()
 
     def run(self, dataset: xr.Dataset, variable_name: str, failures: NDArray[np.bool8]):
-        if self._exceeds_tolerance(failures):
-            raise DataQualityError(
-                f"Quality results for variable {variable_name} indicate a fatal error"
-                " has occurred and the pipeline should exit. Manual review of the data"
-                " is recommended.\n"
-                f"Extra context: '{self.parameters.context}'\n"
-                f"Quality results array: {failures}"
+        if self._exceeds_tolerance(failures):  # default failure tolerance is 0%
+            msg = (
+                f"Quality results for variable '{variable_name}' indicate a fatal error"
+                " has occurred. Manual review of the data is recommended.\n"
             )
 
+            # Show % failed if tolerance is set
+            fail_rate: float = np.average(failures)  # type: ignore
+            msg += (
+                f" {np.count_nonzero(failures)} / {failures.size} values failed"  # type: ignore
+                f" ({100*fail_rate:.2f}%), exceeding the allowable threshold of"
+                f" {100*self.parameters.tolerance}%.\n"
+            )
+
+            # Want to show the first few indexes where the test failed and also the
+            # corresponding data values. Careful to not show too many, otherwise the
+            # message will be bloated and hard to read. Note that np.nonzero(failures)
+            # returns a hard-to-read tuple of indexes, so we modify that to be easier to
+            # read and show the first self.parameters.display_limit # of errors.
+            failed_where = np.nonzero(failures)  # type: ignore
+            failed_values = list(dataset[variable_name].values[failed_where][: self.parameters.display_limit])  # type: ignore
+            failed_indexes: Union[List[int], List[List[int]]]
+            if len(failed_where) == 1:  # 1D
+                failed_indexes = list(failed_where[0][: self.parameters.display_limit])
+            else:
+                failed_indexes = [
+                    [dim_idxs[i] for dim_idxs in failed_where]
+                    for i in range(
+                        min(self.parameters.display_limit, len(failed_where[0]))
+                    )
+                ]
+            msg += (
+                f"The first failures occur at indexes: {failed_indexes}. The"
+                f" corresponding values are: {failed_values}.\n"
+            )
+
+            raise DataQualityError(msg)
         return dataset
 
     def _exceeds_tolerance(self, failures: NDArray[np.bool8]) -> bool:
+        if self.parameters.tolerance == 0:
+            return bool(failures.any())
         failure_ratio: float = np.average(failures)  # type: ignore
         return failure_ratio > self.parameters.tolerance
 

